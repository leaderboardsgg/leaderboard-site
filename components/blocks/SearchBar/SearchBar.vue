<script setup lang="ts">
import type { Ref } from 'vue'
import BaseButton from 'elements/buttons/BaseButton/BaseButton.vue'
import BaseInput from 'elements/inputs/BaseInput/BaseInput.vue'

const emit = defineEmits<{ (e: 'search', value: string): void }>()

interface SearchBarState {
  input: Ref<string>
}

const state: SearchBarState = {
  input: ref(''),
}

function onClick() {
  emit('search', state.input.value)
}

function onEnter() {
  emit('search', state.input.value)
}
</script>

<template>
  <div class="search-bar">
    <BaseInput
      :model="state.input"
      data-testid="search-input"
      name="search"
      placeholder="search..."
      @keyup.enter="onEnter"
    />
    <BaseButton data-testid="search-button" type="button" @click="onClick">
      <i-svg-search />
    </BaseButton>
  </div>
</template>

<style lang="postcss" scoped>
.search-bar {
  @apply flex items-center px-2 relative w-full;
}

.search-bar > input {
  @apply border border-gray-300 rounded bg-white w-full pr-12 px-4 py-1 focus:outline-none;
}

.search-bar > input::placeholder {
  @apply text-gray-300;
}

.search-bar > button {
  @apply absolute right-4 hover:bg-transparent;
}

.search-bar > button > svg {
  @apply h-5 w-5 stroke-gray-300 hover:stroke-black;
}
</style>
