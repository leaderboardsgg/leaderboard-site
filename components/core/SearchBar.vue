<template>
  <div
    class="search-bar"
  >
    <input
      name="search"
      placeholder="search..."
    />
    <button type="submit">
      <svg
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 24 24"
        stroke="currentColor"
      >
        <path
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
        />
      </svg>
    </button>
  </div>
</template>

<style scoped>
  @layer components {
    .search-bar {
      @apply flex items-center px-2 relative text-gray-300 w-full;
    }

    .search-bar > input {
      @apply border border-current rounded bg-white w-full pr-12 px-4 py-1 focus:outline-none;
    }

    .search-bar > button {
      @apply absolute right-4 text-current;
    }

    .search-bar > button > svg {
      @apply h-5 w-5 stroke-current;
    }
  }
</style>
