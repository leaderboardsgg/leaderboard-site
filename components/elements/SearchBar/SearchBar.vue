<template>
  <div class="search-bar">
    <input
      data-testid="search-input"
      name="search"
      placeholder="search..."
      @keyup.enter="onEnter"
    />
    <button data-testid="search-button" type="submit" @click="onClick">
      <svg-icon name="search" />
    </button>
  </div>
</template>

<script lang="ts">
import { defineComponent } from '@nuxtjs/composition-api';

export default defineComponent<{ onSubmit(): void }>({
  name: 'SearchBar',
  props: {
    onSubmit: {
      default: () => console.log('submit'), // eslint-disable-line no-console
      type: Function,
    },
  },
  methods: {
    onClick() {
      this.$props.onSubmit();
    },
    onEnter() {
      this.$props.onSubmit();
    },
  },
});
</script>

<style scoped>
.search-bar {
  @apply flex items-center px-2 relative text-gray-300 w-full;
}

.search-bar > input {
  @apply border border-current rounded bg-white w-full pr-12 px-4 py-1 focus:outline-none;
}

.search-bar > button {
  @apply absolute right-4 text-current;
}

.search-bar > button > svg {
  @apply h-5 w-5 stroke-current;
}
</style>
