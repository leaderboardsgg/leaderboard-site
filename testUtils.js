import Vue from 'vue';
import { RouterLinkStub } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/vue';

export const NuxtLinkStub = Vue.component('NuxtLinkStub', {
  extends: RouterLinkStub,
  render(h) {
    const temp = h(
      this.tag || 'a',
      {
        attrs: {
          ...this.$vnode.data.attrs,
          ...this.$attrs,
          href: this.to,
        },
      },
      this.$slots.default,
    );

    return temp;
  },
});

export const stubbedRender = (Component, options) => {
  return render(Component, {
    stubs: { NuxtLink: NuxtLinkStub, ...options?.stubs },
    ...options,
  });
};

export * from '@testing-library/vue';
export * from '@vue/test-utils';
export { userEvent as fireEvent };
