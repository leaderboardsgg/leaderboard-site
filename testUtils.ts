import { createElementBlock, Component, ComponentOptions } from 'vue'
import { MountingOptions } from '@vue/test-utils'
import userEvent from '@testing-library/user-event'
import { render, RenderOptions } from '@testing-library/vue'

export const stubbedRender = (
  Component: Component | ComponentOptions,
  options?: RenderOptions | MountingOptions<any>,
) => {
  return render(Component, {
    global: {
      stubs: {
        NuxtLink: createElementBlock('a'),
      },
    },
    ...options,
  })
}

export * from '@testing-library/vue'
export * from '@nuxt/test-utils'
export * from '@vue/test-utils'
export { userEvent as fireEvent }
