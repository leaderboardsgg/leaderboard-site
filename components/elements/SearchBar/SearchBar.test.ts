import SearchBar from './SearchBar.vue';
import { fireEvent, stubbedRender } from '@/testUtils';

describe('<SearchBar />', () => {
  it('should render without crashing', () => {
    const { unmount } = stubbedRender(SearchBar);

    unmount();
  });

  it('renders correctly', () => {
    const { container } = stubbedRender(SearchBar);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('when the search event is emitted', () => {
    const inputValue = 'test';

    test('when the search button is clicked', async () => {
      const { emitted, getByTestId } = stubbedRender(SearchBar);

      await fireEvent.type(getByTestId('search-input'), inputValue);
      await fireEvent.click(getByTestId('search-button'));

      expect(emitted().search).toBeTruthy();
      expect(emitted().search?.[0]?.[0]).toEqual(inputValue);
    });

    test('when the enter key is released', async () => {
      const { emitted, getByTestId } = stubbedRender(SearchBar);

      await fireEvent.type(getByTestId('search-input'), inputValue);
      await fireEvent.type(getByTestId('search-input'), '{enter}');

      expect(emitted().search).toBeTruthy();
      expect(emitted().search?.[0]?.[0]).toEqual(inputValue);
    });
  });
});
