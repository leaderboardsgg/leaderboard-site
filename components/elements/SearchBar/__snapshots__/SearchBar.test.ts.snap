// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> renders correctly 1`] = `
<div
  class="search-bar"
>
  <input
    data-testid="search-input"
    name="search"
    placeholder="search..."
  />
   
  <button
    data-testid="search-button"
    type="submit"
  >
    <svg-icon-stub
      name="search"
    />
  </button>
</div>
`;
